File: cpp/gas.cpp
#include "gas.h"

HydroScratchPad::HydroScratchPad( int mesh_size ) :
    density( mesh_size ), mom_n( mesh_size ), mom_t1( mesh_size ), mom_t2( mesh_size ),
    energy( mesh_size ), v_n( mesh_size ), v_t1( mesh_size ), v_t2( mesh_size ),
    pressure( mesh_size ), rho_L( mesh_size ), p_L( mesh_size ),
    vn_L( mesh_size ), vt1_L( mesh_size ), vt2_L( mesh_size ), E_L( mesh_size ),
    mom_n_L( mesh_size ), mom_t1_L( mesh_size ), mom_t2_L( mesh_size ), rho_R( mesh_size ),
    p_R( mesh_size ), vn_R( mesh_size ), vt1_R( mesh_size ), vt2_R( mesh_size ),
    E_R( mesh_size ), mom_n_R( mesh_size ), mom_t1_R( mesh_size ), mom_t2_R( mesh_size ),
    cs_L( mesh_size ), cs_R( mesh_size ), S_L( mesh_size ), S_R( mesh_size ),
    S_R_minus_S_L( mesh_size ), F_dens_L( mesh_size ), F_dens_R( mesh_size ),
    F_momn_L( mesh_size ), F_momn_R( mesh_size ), F_momt1_L( mesh_size ),
    F_momt1_R( mesh_size ), F_momt2_L( mesh_size ), F_momt2_R( mesh_size ),
    F_en_L( mesh_size ), F_en_R( mesh_size ),
    flux_density( mesh_size ), flux_mom_n( mesh_size ),
    flux_mom_t1( mesh_size ), flux_mom_t2( mesh_size ), flux_energy( mesh_size ) {
}

Grid3D roll( const Grid3D& m, int shift, int axis ) {
    int N = m.n;
    Grid3D res( N );
    for( int i = 0; i < N; ++i ) {
        for( int j = 0; j < N; ++j ) {
            for( int k = 0; k < N; ++k ) {
                if( axis == 0 ) res( ( i + shift + N ) % N, j, k ) = m( i, j, k );
                else if( axis == 1 ) res( i, ( j + shift + N ) % N, k ) = m( i, j, k );
                else if( axis == 2 ) res( i, j, ( k + shift + N ) % N ) = m( i, j, k );
            }
        }
    }
    return res;
}

GasGrid::GasGrid( const Config& conf ) :
    density( conf.MESH_SIZE ), momentum_x( conf.MESH_SIZE ), momentum_y( conf.MESH_SIZE ), momentum_z( conf.MESH_SIZE ),
    energy( conf.MESH_SIZE ), pressure( conf.MESH_SIZE ), velocity_x( conf.MESH_SIZE ), velocity_y( conf.MESH_SIZE ), velocity_z( conf.MESH_SIZE ),
    accel_x( conf.MESH_SIZE ), accel_y( conf.MESH_SIZE ), accel_z( conf.MESH_SIZE ),
    scratch( conf.MESH_SIZE ), config( conf )
{
}

void GasGrid::update_primitive_variables() {
    for( int i = 0; i < config.MESH_SIZE * config.MESH_SIZE * config.MESH_SIZE; ++i ) {
        if( density.data[i] > 1e-12 ) {
            velocity_x.data[i] = momentum_x.data[i] / density.data[i];
            velocity_y.data[i] = momentum_y.data[i] / density.data[i];
            velocity_z.data[i] = momentum_z.data[i] / density.data[i];
        }
        else {
            velocity_x.data[i] = velocity_y.data[i] = velocity_z.data[i] = 0.0;
        }
    }

    Grid3D kin_energy( config.MESH_SIZE );
    kin_energy.data = 0.5 * ( momentum_x.array().square() + momentum_y.array().square() + momentum_z.array().square() );

    for( int i = 0; i < config.MESH_SIZE * config.MESH_SIZE * config.MESH_SIZE; ++i ) {
        if( density.data[i] > 1e-12 ) kin_energy.data[i] /= density.data[i];
        else kin_energy.data[i] = 0.0;
    }

    pressure.data = ( config.GAMMA - 1.0 ) * ( energy.data - kin_energy.data );
    pressure.data = ( pressure.array() < 1e-12 ).select( 1e-12, pressure.data );
}

void GasGrid::calculate_fluxes( int axis ) {
    scratch.density = density;
    scratch.energy = energy;
    scratch.pressure = pressure;

    if( axis == 0 ) {
        scratch.mom_n = momentum_x; scratch.mom_t1 = momentum_y; scratch.mom_t2 = momentum_z;
        scratch.v_n = velocity_x; scratch.v_t1 = velocity_y; scratch.v_t2 = velocity_z;
    }
    else if( axis == 1 ) {
        scratch.mom_n = momentum_y; scratch.mom_t1 = momentum_z; scratch.mom_t2 = momentum_x;
        scratch.v_n = velocity_y; scratch.v_t1 = velocity_z; scratch.v_t2 = velocity_x;
    }
    else {
        scratch.mom_n = momentum_z; scratch.mom_t1 = momentum_x; scratch.mom_t2 = momentum_y;
        scratch.v_n = velocity_z; scratch.v_t1 = velocity_x; scratch.v_t2 = velocity_y;
    }

    const int rollDir = -1;
    scratch.rho_L = scratch.density;
    scratch.p_L = scratch.pressure;
    scratch.vn_L = scratch.v_n;
    scratch.vt1_L = scratch.v_t1;
    scratch.vt2_L = scratch.v_t2;
    scratch.E_L = scratch.energy;
    scratch.mom_n_L = scratch.mom_n;
    scratch.mom_t1_L = scratch.mom_t1;
    scratch.mom_t2_L = scratch.mom_t2;

    scratch.rho_R = roll( scratch.rho_L, rollDir, axis );
    scratch.p_R = roll( scratch.p_L, rollDir, axis );
    scratch.vn_R = roll( scratch.vn_L, rollDir, axis );
    scratch.vt1_R = roll( scratch.vt1_L, rollDir, axis );
    scratch.vt2_R = roll( scratch.vt2_L, rollDir, axis );
    scratch.E_R = roll( scratch.E_L, rollDir, axis );
    scratch.mom_n_R = roll( scratch.mom_n_L, rollDir, axis );
    scratch.mom_t1_R = roll( scratch.mom_t1_L, rollDir, axis );
    scratch.mom_t2_R = roll( scratch.mom_t2_L, rollDir, axis );

    scratch.cs_L.data = ( config.GAMMA * scratch.p_L.array() / scratch.rho_L.array() ).sqrt();
    scratch.cs_R.data = ( config.GAMMA * scratch.p_R.array() / scratch.rho_R.array() ).sqrt();
    scratch.cs_L.data = ( scratch.rho_L.array() > 1e-12 ).select( scratch.cs_L.array(), 0.0 );
    scratch.cs_R.data = ( scratch.rho_R.array() > 1e-12 ).select( scratch.cs_R.array(), 0.0 );
    scratch.cs_L.data = ( scratch.cs_L.array() == scratch.cs_L.array() ).select( scratch.cs_L.array(), 0.0 );
    scratch.cs_R.data = ( scratch.cs_R.array() == scratch.cs_R.array() ).select( scratch.cs_R.array(), 0.0 );

    scratch.S_L.data = ( scratch.vn_L.array() - scratch.cs_L.array() ).cwiseMin( scratch.vn_R.array() - scratch.cs_R.array() );
    scratch.S_R.data = ( scratch.vn_L.array() + scratch.cs_L.array() ).cwiseMax( scratch.vn_R.array() + scratch.cs_R.array() );

    scratch.F_dens_L.data = scratch.rho_L.array() * scratch.vn_L.array();
    scratch.F_dens_R.data = scratch.rho_R.array() * scratch.vn_R.array();
    scratch.F_momn_L.data = scratch.rho_L.array() * scratch.vn_L.array().square() + scratch.p_L.array();
    scratch.F_momn_R.data = scratch.rho_R.array() * scratch.vn_R.array().square() + scratch.p_R.array();
    scratch.F_momt1_L.data = scratch.rho_L.array() * scratch.vn_L.array() * scratch.vt1_L.array();
    scratch.F_momt1_R.data = scratch.rho_R.array() * scratch.vn_R.array() * scratch.vt1_R.array();
    scratch.F_momt2_L.data = scratch.rho_L.array() * scratch.vn_L.array() * scratch.vt2_L.array();
    scratch.F_momt2_R.data = scratch.rho_R.array() * scratch.vn_R.array() * scratch.vt2_R.array();
    scratch.F_en_L.data = ( scratch.E_L.array() + scratch.p_L.array() ) * scratch.vn_L.array();
    scratch.F_en_R.data = ( scratch.E_R.array() + scratch.p_R.array() ) * scratch.vn_R.array();

    scratch.S_R_minus_S_L.data = scratch.S_R.array() - scratch.S_L.array();
    scratch.S_R_minus_S_L.data = ( scratch.S_R_minus_S_L.array().abs() < 1e-9 ).select( 1e-9, scratch.S_R_minus_S_L.data );

    auto HLL = [&]( const Grid3D& FL, const Grid3D& FR, const Grid3D& UL, const Grid3D& UR ) {
        Grid3D flux( config.MESH_SIZE );
        flux.data = ( scratch.S_L.array() >= 0 ).select( FL.array(),
            ( scratch.S_R.array() <= 0 ).select( FR.array(),
                ( scratch.S_R.array() * FL.array() - scratch.S_L.array() * FR.array() +
                    scratch.S_L.array() * scratch.S_R.array() * ( UR.array() - UL.array() ) ) / scratch.S_R_minus_S_L.array() ) );
        return flux;
        };

    scratch.flux_density = HLL( scratch.F_dens_L, scratch.F_dens_R, scratch.rho_L, scratch.rho_R );
    scratch.flux_mom_n = HLL( scratch.F_momn_L, scratch.F_momn_R, scratch.mom_n_L, scratch.mom_n_R );
    scratch.flux_mom_t1 = HLL( scratch.F_momt1_L, scratch.F_momt1_R, scratch.mom_t1_L, scratch.mom_t1_R );
    scratch.flux_mom_t2 = HLL( scratch.F_momt2_L, scratch.F_momt2_R, scratch.mom_t2_L, scratch.mom_t2_R );
    scratch.flux_energy = HLL( scratch.F_en_L, scratch.F_en_R, scratch.E_L, scratch.E_R );
}

void GasGrid::hydro_step( double dt ) {
    update_primitive_variables();
    double factor = dt / config.CELL_SIZE;

    // X-Sweep
    calculate_fluxes( 0 );
    density.data = density.array() - factor * ( scratch.flux_density.array() - roll( scratch.flux_density, 1, 0 ).array() );
    momentum_x.data = momentum_x.array() - factor * ( scratch.flux_mom_n.array() - roll( scratch.flux_mom_n, 1, 0 ).array() );
    momentum_y.data = momentum_y.array() - factor * ( scratch.flux_mom_t1.array() - roll( scratch.flux_mom_t1, 1, 0 ).array() );
    momentum_z.data = momentum_z.array() - factor * ( scratch.flux_mom_t2.array() - roll( scratch.flux_mom_t2, 1, 0 ).array() );
    energy.data = energy.array() - factor * ( scratch.flux_energy.array() - roll( scratch.flux_energy, 1, 0 ).array() );
    update_primitive_variables();

    // Y-Sweep
    calculate_fluxes( 1 );
    density.data = density.array() - factor * ( scratch.flux_density.array() - roll( scratch.flux_density, 1, 1 ).array() );
    momentum_x.data = momentum_x.array() - factor * ( scratch.flux_mom_t2.array() - roll( scratch.flux_mom_t2, 1, 1 ).array() );
    momentum_y.data = momentum_y.array() - factor * ( scratch.flux_mom_n.array() - roll( scratch.flux_mom_n, 1, 1 ).array() );
    momentum_z.data = momentum_z.array() - factor * ( scratch.flux_mom_t1.array() - roll( scratch.flux_mom_t1, 1, 1 ).array() );
    energy.data = energy.array() - factor * ( scratch.flux_energy.array() - roll( scratch.flux_energy, 1, 1 ).array() );
    update_primitive_variables();

    // Z-Sweep
    calculate_fluxes( 2 );
    density.data = density.array() - factor * ( scratch.flux_density.array() - roll( scratch.flux_density, 1, 2 ).array() );
    momentum_x.data = momentum_x.array() - factor * ( scratch.flux_mom_t1.array() - roll( scratch.flux_mom_t1, 1, 2 ).array() );
    momentum_y.data = momentum_y.array() - factor * ( scratch.flux_mom_t2.array() - roll( scratch.flux_mom_t2, 1, 2 ).array() );
    momentum_z.data = momentum_z.array() - factor * ( scratch.flux_mom_n.array() - roll( scratch.flux_mom_n, 1, 2 ).array() );
    energy.data = energy.array() - factor * ( scratch.flux_energy.array() - roll( scratch.flux_energy, 1, 2 ).array() );
    update_primitive_variables();
}

double GasGrid::get_cfl_timestep() const {
    if( !config.USE_HYDRO ) return std::numeric_limits<double>::infinity();

    Grid3D cs_sq( config.MESH_SIZE );
    cs_sq.data = ( config.GAMMA * pressure.array() / density.array() );
    cs_sq.data = ( density.array() > 1e-12 ).select( cs_sq.array(), 0.0 );
    cs_sq.data = ( cs_sq.array() == cs_sq.array() ).select( cs_sq.array(), 0.0 );

    Grid3D v_mag( config.MESH_SIZE );
    v_mag.data = ( velocity_x.array().square() + velocity_y.array().square() + velocity_z.array().square() ).sqrt();

    Grid3D signal_vel( config.MESH_SIZE );
    signal_vel.data = v_mag.array() + cs_sq.array().sqrt();

    double max_signal_vel = signal_vel.maxCoeff();
    if( max_signal_vel < 1e-9 ) return std::numeric_limits<double>::infinity();

    return ( config.CELL_SIZE / max_signal_vel ) * config.CFL_SAFETY_FACTOR;
}

File: python/gas.py
import numpy as np

class GasGrid:
    def __init__(self, config):
        self.config = config
        mesh_size = config.mesh_size
        grid_shape = (mesh_size, mesh_size, mesh_size)
        
        self.density = np.zeros(grid_shape)
        self.momentum_x = np.zeros(grid_shape)
        self.momentum_y = np.zeros(grid_shape)
        self.momentum_z = np.zeros(grid_shape)
        self.energy = np.zeros(grid_shape)
        
        self.pressure = np.zeros(grid_shape)
        self.velocity_x = np.zeros(grid_shape)
        self.velocity_y = np.zeros(grid_shape)
        self.velocity_z = np.zeros(grid_shape)

        self.accel_x = np.zeros(grid_shape)
        self.accel_y = np.zeros(grid_shape)
        self.accel_z = np.zeros(grid_shape)

    def update_primitive_variables(self):
        non_zero_density = self.density > 1e-12
        self.velocity_x.fill(0.0)
        self.velocity_y.fill(0.0)
        self.velocity_z.fill(0.0)
        
        self.velocity_x[non_zero_density] = self.momentum_x[non_zero_density] / self.density[non_zero_density]
        self.velocity_y[non_zero_density] = self.momentum_y[non_zero_density] / self.density[non_zero_density]
        self.velocity_z[non_zero_density] = self.momentum_z[non_zero_density] / self.density[non_zero_density]

        kinetic_energy_density = np.zeros_like(self.density)
        kinetic_energy_density[non_zero_density] = 0.5 * (
            self.momentum_x[non_zero_density]**2 + 
            self.momentum_y[non_zero_density]**2 +
            self.momentum_z[non_zero_density]**2
        ) / self.density[non_zero_density]
        
        internal_energy_density = self.energy - kinetic_energy_density
        self.pressure = (self.config.gamma - 1.0) * internal_energy_density
        self.pressure[self.pressure < 1e-12] = 1e-12

    def calculate_fluxes(self, axis):
        if axis == 0: 
            density, mom_n, mom_t1, mom_t2, energy, v_n, v_t1, v_t2, pressure = \
                self.density, self.momentum_x, self.momentum_y, self.momentum_z, self.energy, \
                self.velocity_x, self.velocity_y, self.velocity_z, self.pressure
        elif axis == 1: 
            density = np.transpose(self.density, (1, 0, 2))
            mom_n = np.transpose(self.momentum_y, (1, 0, 2))
            mom_t1 = np.transpose(self.momentum_x, (1, 0, 2))
            mom_t2 = np.transpose(self.momentum_z, (1, 0, 2))
            energy = np.transpose(self.energy, (1, 0, 2))
            v_n = np.transpose(self.velocity_y, (1, 0, 2))
            v_t1 = np.transpose(self.velocity_x, (1, 0, 2))
            v_t2 = np.transpose(self.velocity_z, (1, 0, 2))
            pressure = np.transpose(self.pressure, (1, 0, 2))
        elif axis == 2: 
            density = np.transpose(self.density, (2, 0, 1))
            mom_n = np.transpose(self.momentum_z, (2, 0, 1))
            mom_t1 = np.transpose(self.momentum_x, (2, 0, 1))
            mom_t2 = np.transpose(self.momentum_y, (2, 0, 1))
            energy = np.transpose(self.energy, (2, 0, 1))
            v_n = np.transpose(self.velocity_z, (2, 0, 1))
            v_t1 = np.transpose(self.velocity_x, (2, 0, 1))
            v_t2 = np.transpose(self.velocity_y, (2, 0, 1))
            pressure = np.transpose(self.pressure, (2, 0, 1))

        rho_L, p_L, vn_L, vt1_L, vt2_L, E_L, mom_n_L, mom_t1_L, mom_t2_L = \
            density, pressure, v_n, v_t1, v_t2, energy, mom_n, mom_t1, mom_t2
        
        rho_R, p_R, vn_R, vt1_R, vt2_R, E_R, mom_n_R, mom_t1_R, mom_t2_R = \
            np.roll(rho_L, -1, axis=0), np.roll(p_L, -1, axis=0), np.roll(vn_L, -1, axis=0), \
            np.roll(vt1_L, -1, axis=0), np.roll(vt2_L, -1, axis=0), np.roll(E_L, -1, axis=0), \
            np.roll(mom_n_L, -1, axis=0), np.roll(mom_t1_L, -1, axis=0), np.roll(mom_t2_L, -1, axis=0)

        gamma = self.config.gamma
        cs_L = np.sqrt(gamma * p_L / rho_L, where=rho_L > 1e-12, out=np.zeros_like(rho_L))
        cs_R = np.sqrt(gamma * p_R / rho_R, where=rho_R > 1e-12, out=np.zeros_like(rho_R))

        S_L = np.minimum(vn_L - cs_L, vn_R - cs_R)
        S_R = np.maximum(vn_L + cs_L, vn_R + cs_R)

        F_dens_L, F_dens_R = rho_L * vn_L, rho_R * vn_R
        F_momn_L, F_momn_R = rho_L * vn_L**2 + p_L, rho_R * vn_R**2 + p_R
        F_momt1_L, F_momt1_R = rho_L * vn_L * vt1_L, rho_R * vn_R * vt1_R
        F_momt2_L, F_momt2_R = rho_L * vn_L * vt2_L, rho_R * vn_R * vt2_R
        F_en_L, F_en_R = (E_L + p_L) * vn_L, (E_R + p_R) * vn_R

        S_R_minus_S_L = S_R - S_L
        S_R_minus_S_L[np.abs(S_R_minus_S_L) < 1e-9] = 1e-9 

        flux_density = np.where(S_L >= 0, F_dens_L, 
                    np.where(S_R <= 0, F_dens_R, 
                                (S_R*F_dens_L - S_L*F_dens_R + S_L*S_R*(rho_R-rho_L))/S_R_minus_S_L))
        flux_mom_n = np.where(S_L >= 0, F_momn_L,
                    np.where(S_R <= 0, F_momn_R,
                            (S_R*F_momn_L - S_L*F_momn_R + S_L*S_R*(mom_n_R-mom_n_L))/S_R_minus_S_L))
        flux_mom_t1 = np.where(S_L >= 0, F_momt1_L,
                    np.where(S_R <= 0, F_momt1_R,
                            (S_R*F_momt1_L - S_L*F_momt1_R + S_L*S_R*(mom_t1_R-mom_t1_L))/S_R_minus_S_L))
        flux_mom_t2 = np.where(S_L >= 0, F_momt2_L,
                    np.where(S_R <= 0, F_momt2_R,
                            (S_R*F_momt2_L - S_L*F_momt2_R + S_L*S_R*(mom_t2_R-mom_t2_L))/S_R_minus_S_L))
        flux_energy = np.where(S_L >= 0, F_en_L,
                    np.where(S_R <= 0, F_en_R,
                            (S_R*F_en_L - S_L*F_en_R + S_L*S_R*(E_R-E_L))/S_R_minus_S_L))

        if axis == 0: 
            return flux_density, flux_mom_n, flux_mom_t1, flux_mom_t2, flux_energy
        elif axis == 1: 
            return flux_density.transpose(1, 0, 2), flux_mom_t1.transpose(1, 0, 2), \
                   flux_mom_n.transpose(1, 0, 2), flux_mom_t2.transpose(1, 0, 2), flux_energy.transpose(1, 0, 2)
        elif axis == 2: 
            return flux_density.transpose(1, 2, 0), flux_mom_t1.transpose(1, 2, 0), \
                   flux_mom_t2.transpose(1, 2, 0), flux_mom_n.transpose(1, 2, 0), flux_energy.transpose(1, 2, 0)

    def hydro_step(self, dt):
        self.update_primitive_variables()
        factor = dt / self.config.cell_size
        
        flux_density, flux_mom_x, flux_mom_y, flux_mom_z, flux_energy = self.calculate_fluxes(axis=0)
        self.density    -= factor * (flux_density - np.roll(flux_density, 1, axis=0))
        self.momentum_x -= factor * (flux_mom_x - np.roll(flux_mom_x, 1, axis=0))
        self.momentum_y -= factor * (flux_mom_y - np.roll(flux_mom_y, 1, axis=0))
        self.momentum_z -= factor * (flux_mom_z - np.roll(flux_mom_z, 1, axis=0))
        self.energy     -= factor * (flux_energy - np.roll(flux_energy, 1, axis=0))
        self.update_primitive_variables()

        flux_density, flux_mom_x, flux_mom_y, flux_mom_z, flux_energy = self.calculate_fluxes(axis=1)
        self.density    -= factor * (flux_density - np.roll(flux_density, 1, axis=1))
        self.momentum_x -= factor * (flux_mom_x - np.roll(flux_mom_x, 1, axis=1))
        self.momentum_y -= factor * (flux_mom_y - np.roll(flux_mom_y, 1, axis=1))
        self.momentum_z -= factor * (flux_mom_z - np.roll(flux_mom_z, 1, axis=1))
        self.energy     -= factor * (flux_energy - np.roll(flux_energy, 1, axis=1))
        self.update_primitive_variables()
        
        flux_density, flux_mom_x, flux_mom_y, flux_mom_z, flux_energy = self.calculate_fluxes(axis=2)
        self.density    -= factor * (flux_density - np.roll(flux_density, 1, axis=2))
        self.momentum_x -= factor * (flux_mom_x - np.roll(flux_mom_x, 1, axis=2))
        self.momentum_y -= factor * (flux_mom_y - np.roll(flux_mom_y, 1, axis=2))
        self.momentum_z -= factor * (flux_mom_z - np.roll(flux_mom_z, 1, axis=2))
        self.energy     -= factor * (flux_energy - np.roll(flux_energy, 1, axis=2))
        self.update_primitive_variables()

def get_cfl_timestep(gas, config):
    if not config.enable_hydro or not np.any(gas.density > 1e-12): return np.inf
    non_zero = gas.density > 1e-12
    cs = np.sqrt(config.gamma * gas.pressure[non_zero] / gas.density[non_zero])
    v_mag = np.sqrt(gas.velocity_x[non_zero]**2 + gas.velocity_y[non_zero]**2 + gas.velocity_z[non_zero]**2)
    max_signal = np.max(v_mag + cs)
    return config.cell_size / max_signal * config.cfl_safety_factor if max_signal >= 1e-9 else np.inf